[css-anchor-position-1] PositionOptions::currentOption should always set lastSuccessfulPositionTryFallbackIndex

tuankiet65 opened the pull request:
#### eb8b277b0314108e5cb2886e39dfc52d98fc845a
<pre>
[css-anchor-position-1] PositionOptions::currentOption should always set lastSuccessfulPositionTryFallbackIndex
<a href="https://rdar.apple.com/158894957">rdar://158894957</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=297754">https://bugs.webkit.org/show_bug.cgi?id=297754</a>

Reviewed by Antti Koivisto.

TreeResolver::PositionOptions::currentOption() sets
lastSuccessfulPositionTryFallbackIndex of the cloned style to be the index
of the current option, except when the current option doesn&apos;t have an index
(the index is nullopt.) The except part is wrong, because a nullopt current
option index means it&apos;s the base style, so lastSuccessfulPositionTryFallbackIndex
should also be set to nullopt, to indicate the last-successful position option
is the base style.

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/last-successful-fallback-to-base-style-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/last-successful-fallback-to-base-style.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/last-successful-intermediate-ignored-expected.txt:
    - Rebaseline test that incidentally passes because of the bug.

* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::setLastSuccessfulPositionTryFallbackIndex):
* Source/WebCore/rendering/style/RenderStyle.h:
    - Pass index by value to RenderStyle::setLastSuccessfulPositionTryFallbackIndex.

* Source/WebCore/style/StyleTreeResolver.cpp:
(WebCore::Style::TreeResolver::PositionOptions::currentOption const):

Canonical link: <a href="https://commits.webkit.org/299350@main">https://commits.webkit.org/299350@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4325b74f107e3f96c9a0ae76f216a15a208fb9a6

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/118476 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/38157 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/28808 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/124648 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/70535 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/cb663b8c-fc0c-457a-9611-ff071babf781) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/120354 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/38853 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/46739 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/89925 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/59502 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/91905306-72cc-4db7-9341-abdc0103bd0b) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/121429 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/30955 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/106230 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/70405 "1 api test failed or timed out") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/316f6195-d4be-4949-8d5f-63496e60a88c) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/30013 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/24341 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/68307 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/100392 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/24530 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/127714 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/45383 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/34241 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/98585 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/45747 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/102449 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/98370 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25047 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/43786 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/21778 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/41881 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/45253 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/50931 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/44716 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/48063 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/46403 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

anttijk commented on Source/WebCore/style/StyleTreeResolver.cpp:
```diff
@@ -1566,8 +1566,8 @@ std::unique_ptr<RenderStyle> TreeResolver::PositionOptions::currentOption() cons
-    if (optionStyles[index].fallbackIndex)
-        newStyle->setLastSuccessfulPositionTryFallbackIndex(*optionStyles[index].fallbackIndex);
+    auto fallbackIndex = optionStyles[index].fallbackIndex;
+    newStyle->setLastSuccessfulPositionTryFallbackIndex(WTFMove(fallbackIndex));
```
It is a bit odd factoring/naming that a function called currentOption() (which sounds like a getter) also sets things.

tuankiet65 replied:
This patch https://github.com/WebKit/WebKit/pull/50010 will separate the "used position option index" and "last successful position option index", so in that patch we'll set the "used position option index" at option generation time, so `::currentOption()` won't set it anymore

anttijk commented on LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/last-successful-intermediate-ignored-expected.txt:
```diff
@@ -1,4 +1,4 @@
 
 PASS Starts rendering with flip-block
-PASS No successful position (with intermediate successful), keep flip-block
+FAIL No successful position (with intermediate successful), keep flip-block assert_equals: expected 200 but got 0
```
Why does this regress? Are you fixing it?

tuankiet65 replied:
I haven't looked into it, but I strongly believe it passes on accident, so it's just coincidence that it passes with the last-successful position option patch

anttijk replied:
You should look into it before landing this.

tuankiet65 replied:
The failure is because we've never implemented remembering last successful position option at `ResizeObserver` time. It's implemented here #50010, so I'd like to get this patch in first and #50010 in later.

anttijk reviewed (changes requested):
we should understand the failing test

anttijk reviewed (approved):
ok!

tuankiet65 committed:
[css-anchor-position-1] PositionOptions::currentOption should always set lastSuccessfulPositionTryFallbackIndex
rdar://158894957
https://bugs.webkit.org/show_bug.cgi?id=297754

Reviewed by Antti Koivisto.

TreeResolver::PositionOptions::currentOption() sets
lastSuccessfulPositionTryFallbackIndex of the cloned style to be the index
of the current option, except when the current option doesn't have an index
(the index is nullopt.) The except part is wrong, because a nullopt current
option index means it's the base style, so lastSuccessfulPositionTryFallbackIndex
should also be set to nullopt, to indicate the last-successful position option
is the base style.

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/last-successful-fallback-to-base-style-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/last-successful-fallback-to-base-style.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/last-successful-intermediate-ignored-expected.txt:
    - Rebaseline test that incidentally passes because of the bug.

* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::setLastSuccessfulPositionTryFallbackIndex):
* Source/WebCore/rendering/style/RenderStyle.h:
    - Pass index by value to RenderStyle::setLastSuccessfulPositionTryFallbackIndex.

* Source/WebCore/style/StyleTreeResolver.cpp:
(WebCore::Style::TreeResolver::PositionOptions::currentOption const):

Canonical link: https://commits.webkit.org/299350@main
